Fix default command when hovering objectified units

efrec opened the pull request:
The return value selected with `select(4, Spring.GetUnitHealth(id))` is the captureProgress, so hovering over a fully-captured unit will convert the default command to a CMD_MOVE. I'm not sure units can remain at 100% capture for any time period, so this likely never happens.

This should be `select(5, Spring.GetUnitHealth(id))`, which selects the buildProgress.

Idk what issues this might be causing, if even any. I just noticed this but didn't test.

sprunk commented:
Looks reasonable

WatchTheFort commented:
Is there a way to test this?

saurtron commented:
> > For other units default is to guard them unless damaged, where it will be repair.
> > Just move probably fine there anyways, but imo repair would be perfect when damaged.
> 
> BAR walls have `repairable = false` by convention (not enforced), so this could be improved but isn't an issue currently.
> 
> > Also (I think) managed to test decorations, apparently units like `dice` or `xmasball2_1` have this. Also 'unit hats'. Tried with dice and xmasballs, and they don't seem to trigger `gadget:DefaultCommand`, so likely the `isDecoration[uDefID]` part doesn't do anything there.
> > Also couldn't execute capture on them, or any other commands. So checking capture, or probably build, wouldn't make much sense afaics.
> 
> The decoration units also have `reclaimable = false` and etc. (again, by convention), so that is engine behavior.
> 
> I checked the rest of the objectify code. It does want to disallow attacks, and does so very directly:
> 
> https://github.com/beyond-all-reason/Beyond-All-Reason/blob/dec36a2c9eca8eb92be40156fbb683038340ca5b/luarules/gadgets/unit_objectify.lua#L123-L129
> 
> Which leaves me to wonder how long it's been since I last attacked a wall.

You actually can since not long ago, check latest version:

https://github.com/beyond-all-reason/Beyond-All-Reason/pull/5401/files

efrec commented:
I don't know. I don't think we got the GDT minutes.

FWIW: I think this gadget relies on bad assumptions. We should expect general-purpose code to behave well under modding or, in the case of ceasefire, the mission api.

But probably these should be reviewed rather than spot-checked. I could add an issue to collect bad allegiance checks (I am guilty here), and another to collect weak logical implications that don't hold, like "if objectified then not repairable".

We have related issues, like preventing attack commands snapping to allied units making [allied walls untargetable](https://github.com/beyond-all-reason/Beyond-All-Reason/issues/4468). Of course, there was `CMD_INSERT` trickery to get around even this, but that is [going away forever](https://github.com/beyond-all-reason/Beyond-All-Reason/pull/5560) if I have my way.

PtaQQ commented:
On top of an ally wall: Reclaim -> move (if no reclaim abbility)
On top of an ally wall being built: Repair -> move
On top of an enemy wall: Attack -> reclaim -> move 
(so you can't tell apart armclaw and cormaw from walls)

Official GDT ruling.

efrec commented:
I do not grok that ruling entirely except not to leak intel on decoys (do the arrows mean "replace command with"..?). I replaced mobile units' default commands with Move.

That leaves out immobile units, so I added a default attack command on enemy walls for static defenses.

Unit selections can be large, also, so I need to come back to this and speed it up.

sprunk commented on luarules/gadgets/unit_objectify.lua:
```diff
@@ -151,20 +167,61 @@ if gadgetHandler:IsSyncedCode() then
+				-- This could be more efficient but is not especially costly as-is.
+				for i = 1, 64 do
+					local selectedID = spGetUnitDefID(selected[i])
+					if canMove[selectedID] then
+						return CMD_MOVE
```
wasn't Attack supposed to be higher priority than Move?

sprunk commented on luarules/gadgets/unit_objectify.lua:
```diff
@@ -151,20 +187,81 @@ if gadgetHandler:IsSyncedCode() then
+			for j = 1, #predicates do
+				local predicate = predicates[j]
+				if predicate.check[unitDefID] then
+					if j == 1 then return predicate.command end
```
~~oppa gangnam~~ style
```suggestion
					if j == 1 then
						return predicate.command
					end
```

sprunk commented on luarules/gadgets/unit_objectify.lua:
```diff
@@ -151,20 +187,81 @@ if gadgetHandler:IsSyncedCode() then
-				-- make sure a command given on top of a objectified/decoration unit is a move command
-				if select(4, Spring.GetUnitHealth(id)) == 1 then
-					return CMD_MOVE
+	local function scanSelection(predicates, limit)
```
perhaps add `limit = limit or 64` so one doesn't need to pass a limit explicitly?

sprunk commented on luarules/gadgets/unit_dragons_disguise.lua:
```diff
@@ -31,6 +31,14 @@ end
+function gadget:Initialize()
+	for _, unitID in ipairs(Spring.GetAllUnits()) do
+		if not Spring.GetUnitIsBeingBuilt(unitID) then
+			gadget:UnitFinished(unitID, Spring.GetUnitDefID(unitID), 0)
```
if the arg isn't actually used consider just not passing it at all
